pgjdbc/src/main/java/org/postgresql/jdbc/TypeInfoCache.java

```diff
@@ -177,267 +186,602 @@ public synchronized void addDataType(String type, Class<? extends PGobject> klas
-    if (!((BaseStatement) _getTypeInfoStatement)
-        .executeWithFlags(QueryExecutor.QUERY_SUPPRESS_BEGIN)) {
-      throw new PSQLException(GT.tr("No results were returned by the query."), PSQLState.NO_DATA);
+    We may not want to do this check anyway given the type may not exist, in which case we want to
+    return java.sql.Types.OTHER. However, this is what the current behavior is.
+     */
+    ParsedTypeName typeName = ParsedTypeName.fromString(pgTypeName);
+    if (typeName.isArray()) {
+      return Types.ARRAY;
     }
 
-    ResultSet rs = _getTypeInfoStatement.getResultSet();
+    PgType pgType = fetchPgType(pgTypeName);
 
-    Integer type = null;
-    if (rs.next()) {
-      boolean isArray = rs.getBoolean(1);
-      String typtype = rs.getString(2);
-      if (isArray) {
-        type = Types.ARRAY;
-      } else if ("c".equals(typtype)) {
-        type = Types.STRUCT;
-      } else if ("d".equals(typtype)) {
-        type = Types.DISTINCT;
-      } else if ("e".equals(typtype)) {
-        type = Types.VARCHAR;
-      }
+    if (pgType == null) {
+      return Types.OTHER;
     }
 
-    if (type == null) {
-      type = Types.OTHER;
-    }
-    rs.close();
+    return pgType.sqlType();
+  }
 
-    _pgNameToSQLType.put(pgTypeName, type);
-    return type;
+  enum PgTypeResultSetColumn {
+    OID(1), ON_PATH(2), NSPNAME(3), IS_ARRAY(4),
+    TYPNAME(5), TYPELEM(6), TYPTYPE(7), TYPDELIM(8),
+    ELEMENT_TYPNAME(9), ELEMENT_TYPTYPE(10), ELEMENT_TYPDELIM(11),
+    ARRAY_OID(12), ARRAY_TYPNAME(13);
+    final int idx;
+
+    PgTypeResultSetColumn(int idx) {
+      this.idx = idx;
+    }
   }
 
-  private PreparedStatement getOidStatement(String pgTypeName) throws SQLException {
-    boolean isArray = pgTypeName.endsWith("[]");
-    boolean hasQuote = pgTypeName.contains("\"");
-    int dotIndex = pgTypeName.indexOf('.');
+  /*
+  To join element and array types, use (arr.typelem, arr.typinput) = (e.oid, 'array_in'::regproc).
+  Some types, such as box, have non-zero typelem values and are not arrays. One might be tempted to
+  leverage the fact that typlen = -1 (variable length) for arrays, but that will run afoul with
+  element types that have multiple array and vector implementations such as int2, which has _int2 and
+  int2vector. The _int2 type uses array_in and corresponds to int2[] and smallint[], while the
+  int2vector type uses int2vectorin and is used in system tables.
+
+  The pg_type.typarray column, added in 8.3, makes this unnecessary, so the queries can be updated
+  when 8.2 support is dropped. The efficiencies gained in query performance are negligible in
+  comparison to time spent over the wire and the overhead in maintaining a separate code path for 8.2.
+   */
 
-    if (dotIndex == -1 && !hasQuote && !isArray) {
+  private PreparedStatement getOidStatement(ParsedTypeName typeName) throws SQLException {
+    if (typeName.isSimple()) {
       if (_getOidStatementSimple == null) {
```

Comment:
does access to this need to be synchronized?

Reply:
As you rightly pointed out below, these private methods are only called by public method which is already synchronized. I've added comments to this effect.